### PR TITLE
Add compatibility with Sonata4 and Symfony5

### PR DESCRIPTION
**Pull request by VincentLanglet**
## Changelog

```markdown
### Added
- Added compatibility with Symfony 5
- Added compatibility with Twig 3
```

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -27,22 +27,22 @@
...
         "friendsofsymfony/rest-bundle": "<2.3",
         "jms/serializer": "<0.13",
         "nelmio/api-doc-bundle": "<2.13 || >=4.0",
-        "sonata-project/admin-bundle": "<3.91.1",
-        "sonata-project/block-bundle": "<3.19 || >=4.0",
+        "sonata-project/admin-bundle": "<=4.0.0-alpha-1",
+        "sonata-project/block-bundle": "<4.0",
         "sonata-project/core-bundle": "<3.20",
         "sonata-project/doctrine-orm-admin-bundle": "<3.16"
     },
```
Comment (jordisala1991) [inline composer.json:48]: this should be changed too

**Comment by jordisala1991** [comment]
The error you have now is due to configured in memory users, you probably have null there, you should simply remove that lines.

**Comment by VincentLanglet** [comment]
Green tests @jordisala1991 :)

File: src/Controller/CategoryAdminController.php [hunk 2]
```diff
@@ -27,12 +30,16 @@
...
-     *
-     * @return Response
-     */
-    public function listAction(?Request $request = null)
+    public static function getSubscribedServices(): array
+    {
+        return [
+            'sonata.classification.manager.category' => CategoryManagerInterface::class,
+            'sonata.classification.manager.context' => ContextManagerInterface::class,
+            'twig' => Environment::class,
```
Comment (jordisala1991) [inline src/Controller/CategoryAdminController.php:38]: this is probably at parent already

Comment (VincentLanglet) [reply src/Controller/CategoryAdminController.php]: Fixed

**Review by jordisala1991 — DISMISSED** [review]

**Comment by jordisala1991** [comment]
We need the devkit pr and we should be able to go

**Comment by VincentLanglet** [comment]
> We need the devkit pr and we should be able to go

The pr is opened

**Comment by jordisala1991** [comment]
A changelog should be added probably

**Comment by VincentLanglet** [comment]
> A changelog should be added probably

Done

**Review by jordisala1991 — DISMISSED** [review]

File: src/Admin/CategoryAdmin.php [hunk 3]
```diff
@@ -27,31 +27,17 @@ class CategoryAdmin extends ContextAwareAdmin
...
-    public function configureRoutes(RouteCollection $routes): void
+    public function configureRoutes(RouteCollectionInterface $routes): void
     {
         $routes->add('tree', 'tree');
     }
 
-    public function getFormBuilder()
+    /**
+     * @param array<string, mixed> $formOptions
+     */
```
Comment (franmomu) [inline src/Admin/CategoryAdmin.php:37]: Is this necessary? I guess it should be in the interface

File: src/Admin/CollectionAdmin.php [hunk 4]
```diff
@@ -24,26 +24,12 @@ class CollectionAdmin extends ContextAwareAdmin
...
 
-    // NEXT_MAJOR: remove this override
-    protected $formOptions = [
-        'cascade_validation' => true,
-    ];
-
-    public function getFormBuilder()
+    /**
+     * @param array<string, mixed> $formOptions
+     */
```
Comment (franmomu) [inline src/Admin/CollectionAdmin.php:29]: I guess it should be in the interface.

File: src/Admin/Filter/CategoryFilter.php [hunk 5]
```diff
@@ -32,51 +33,43 @@ public function __construct(CategoryManagerInterface $categoryManager)
         $this->categoryManager = $categoryManager;
     }
 
-    public function filter(ProxyQueryInterface $queryBuilder, $alias, $field, $data): void
+    public function filter(ProxyQueryInterface $queryBuilder, string $alias, string $field, FilterData $data): void
     {
-        if (null === $data || !\is_array($data) || !\array_key_exists('value', $data)) {
+        if (!$data->hasValue()) {
```
Comment (franmomu) [inline src/Admin/Filter/CategoryFilter.php:38]:
```suggestion
        if (!$data->hasValue() || null === $data->getValue()) {
```
I guess something like that could be done to simplify the code.

File: src/Admin/Filter/CategoryFilter.php [hunk 6]
```diff
@@ -32,51 +33,43 @@ public function __construct(CategoryManagerInterface $categoryManager)
...
+        $value = $data->getValue();
+        if ($value) {
             $queryBuilder
                 ->andWhere(sprintf('%s.%s = :category', $alias, $field))
-                ->setParameter('category', $data['value']);
+                ->setParameter('category', $value);
         }
 
-        $this->active = null !== $data['value'];
+        $this->setActive($value);
```
Comment (franmomu) [inline src/Admin/Filter/CategoryFilter.php:49]: `$value` looks like it is not a boolean.

File: src/Admin/Filter/CollectionFilter.php [hunk 7]
```diff
@@ -31,51 +32,43 @@ public function __construct(CollectionManagerInterface $collectionManager)
...
+        $value = $data->getValue();
+        if ($value) {
             $queryBuilder
                 ->andWhere(sprintf('%s.%s = :collection', $alias, $field))
-                ->setParameter('collection', $data['value']);
+                ->setParameter('collection', $value);
         }
 
-        $this->active = null !== $data['value'];
+        $this->setActive($value);
```
Comment (franmomu) [inline src/Admin/Filter/CollectionFilter.php:48]: `$value` looks like it is not a boolean.

File: src/Controller/CategoryAdminController.php [hunk 8]
```diff
@@ -27,12 +29,15 @@
  */
 class CategoryAdminController extends Controller
 {
-    /**
-     * @param Request $request
-     *
-     * @return Response
-     */
-    public function listAction(?Request $request = null)
+    public static function getSubscribedServices(): array
```
Comment (franmomu) [inline src/Controller/CategoryAdminController.php:32]: I guess the controller should be defined as a service, something like https://github.com/sonata-project/SonataAdminBundle/blob/3214dec4699374e647e9e172a1043a71c273edad/src/Resources/config/core.php#L126-L129

File: src/DependencyInjection/SonataClassificationExtension.php [hunk 9]
```diff
@@ -110,117 +109,6 @@ public function configureAdmin($config, ContainerBuilder $container): void
         $container->setParameter('sonata.classification.admin.context.translation_domain', $config['admin']['context']['translation']);
     }
 
-    /**
-     * NEXT_MAJOR: Remove this method.
-     */
-    public function registerDoctrineMapping(array $config): void
```
Comment (franmomu) [inline src/DependencyInjection/SonataClassificationExtension.php:116]: I guess there should be a call to this method that should be removed too.

File: tests/Controller/CategoryAdminControllerTest.php [hunk 10]
```diff
@@ -145,49 +144,16 @@ protected function setUp(): void
...
-            });
+        $this->container->set('admin_code', $this->admin);
+        $this->container->set('sonata.admin.pool', $this->pool);
+        $this->container->set('sonata.classification.manager.category', $this->categoryManager);
+        $this->container->set('sonata.classification.manager.context', $this->contextManager);
+        $this->container->set('twig', $twig);
+        $this->container->set('request_stack', $this->requestStack);
 
         // php 5.3
         $tthis = $this;
```
Comment (franmomu) [inline tests/Controller/CategoryAdminControllerTest.php:155]: I guess this can be removed.

**Review by franmomu — DISMISSED** [review]

**Comment by VincentLanglet** [comment]
Done @franmomu

**Review by jordisala1991 — APPROVED** [review]

File: src/Resources/config/controllers.xml [hunk 11]
```diff
@@ -0,0 +1,11 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+    <services>
+        <service public="true" id="sonata.classification.controller.category_admin" class="Sonata\ClassificationBundle\Controller\CategoryAdminController">
+            <tag name="container.service_subscriber"/>
+            <call method="setContainer">
+                <argument type="service" id="Symfony\Component\DependencyInjection\ContainerInterface"/>
```
Comment (franmomu) [inline src/Resources/config/controllers.xml:7]: this should be `Psr\Container\ContainerInterface`, otherwise we are injecting the full container.

Comment (VincentLanglet) [reply src/Resources/config/controllers.xml:7]: Will try: https://github.com/sonata-project/SonataClassificationBundle/pull/680
But seems like it's not so simple

**Pull request merged**